drivers/wifi/esp_at/esp_socket.c

```diff
@@ -192,8 +192,30 @@ void esp_socket_rx(struct esp_socket *sock, struct net_buf *buf,
 	 * net_context and socket mutex claims matches the TX code path. Failure
 	 * to do so can lead to deadlocks.
 	 */
+	/* In the close path, we will meet deadlock in the scenario:
+	 * 1. on_cmd_ipd/esp_socket_rx invokes esp_socket_ref_from_link_id
+	 *    and increments refcount.
+	 * 2. zvfs_close/esp_put locks cond.lock.
+	 * 3. zvfs_close/esp_put waits on sem_free.
+	 * 4. on_cmd_ipd/esp_socket_rx waits on cond.lock before esp_socket_unref.
+	 * 5. sem_free waits on esp_socket_unref for refcount reaching zero.
+	 */
 	if (sock->context->cond.lock) {
-		k_mutex_lock(sock->context->cond.lock, K_FOREVER);
+		atomic_val_t ref;
+		int ret = -EAGAIN;
+
+		/*
+		 * If the socket is closing, we can ignore the packet and won't
+		 * trap in deadlock.
+		 */
+		while ((ref = atomic_get(&sock->refcount)) > 1 && ret == -EAGAIN) {
```

Comment:
The ref variable does not seem to be used for anything, do we really need it?
Why not
```
while (atomic_get(&sock->refcount) > 1 && ret == -EAGAIN) {
```

Reply:
removed `ref` variable